samza-core/src/main/java/org/apache/samza/zk/ZkBarrierForVersionUpgrade.java

```diff
@@ -178,14 +202,15 @@ public void handleDataChange(String dataPath, Object data) {
       if (notAValidEvent())
         return;
 
-      zkUtils.unsubscribeDataChanges(barrierStatePath, this);
-      barrierListenerOptional.ifPresent(
-          zkBarrierListener -> zkBarrierListener.onBarrierStateChanged(barrierVersion, (State) data));
+      State barrierState = (State) data;
+      if (!Objects.equals(barrierState, State.NEW)) {
```

Comment:
Should be explicit in the state-checks here. The states that we are looking for are terminal states: DONE, TIMED_OUT. This will also make code immune to refactors later - for instance, if someone adds another state to the barrier state machine we don't want this code-path to be triggered and prior subscriptions being cancelled.